Consolidate `fee_amount` and `amount_needed`

csralvall opened the pull request:
### Description

Before this commit `fee_amount` and `amount_needed` were passed as independent
parameters. From the perspective of coin selection algorithms, they are always
used jointly for the same purpose, to create a coin selection with a total
effective value greater than it's summed values.

This commit removes the abstraction that the use of the two parameter
introduced by consolidating both into a single parameter, `target_amount`, who
carries their values added up.

Resolves: bitcoindevkit/bdk#641

### Notes to the reviewers

I just updated old tests and didn't create new ones because almost all changes
are renames and "logic changes" (like the addition of the selection fee) are
tested in the modified tests.

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [ ] I've added tests for the new feature
* [x] I've added docs for the new feature
* [x] I've updated `CHANGELOG.md`

#### Bugfixes:

* [x] This pull request breaks the existing API
* [x] I'm linking the issue being fixed by this PR

csralvall committed:
Consolidate `fee_amount` and `amount_needed`

Before this commit `fee_amount` and `amount_needed` were passed as independent
parameters. From the perspective of coin selection algorithms, they are always
used jointly for the same purpose, to create a coin selection with a total
effective value greater than it's summed values.

This commit removes the abstraction that the use of the two parameter
introduced by consolidating both into a single parameter, `target_amount`, who
carries their values added up.

danielabrozzoni commented on src/wallet/coin_selection.rs:
```diff
@@ -1148,23 +1129,23 @@ mod test {
         let input_size = (TXIN_BASE_WEIGHT + P2WPKH_WITNESS_SIZE).vbytes();
-        let epsilon = 0.5;
-        assert!((1.0 - (result.fee_amount as f32 / input_size as f32)).abs() < epsilon);
+        // the final fee rate should be exactly the same that the fee rate given
```
nit: the same "as"

Also, is there a specific reason for changing this line? I think we should be more precise with the feerate and always check that it's >= than what the user asked, but maybe we shouldn't change it in this PR?

csralvall replied:
I thought that as we were using the first utxo's effective value as target amount an from this PR on we don't have to consider the accumulated fee amount of the overall transaction, it was a good way to illustrate the idea that BnB would produce an exact match. I don't have strong opinions on this, so I can roll it back.

afilini replied:
Maybe it's just me but I always try to avoid comparing floats with strict equality because there's inherent rounding happening and I'm always afraid it will cause something to break sooner or later.

Maybe in this case it won't happen, but just to be safe I would keep an epsilon, even if very small.

csralvall replied:
Do you have any suggestion for its value? I think it should be less than 0.5 because we are not considering the overall `fee_amount` of the transaction as part of the `fee_rate` for the selection. But I don't know how to avoid magical numbers and use something more "analytical".

evanlinjin replied:
It does seem that this test should be for "an exact match" whilst taking FP rounding errors into consideration. I also agree that `0.5` is probably too big ([machine epsilon for `binary32` is like 24 decimal places](https://en.wikipedia.org/wiki/Machine_epsilon) which roughly means, based on my limited understanding, numbers being stored with 23-24 digits of precision).

I would say, having `epsilon` as `0.01` should be plenty lol (no point going any smaller imo).

csralvall replied:
LGTM, done!

csralvall replied:
I found there is a constant defined for `f32` machine epsilon, probably for this kind of cases. I'll use it instead of `0.01`.

danielabrozzoni reviewed (commented):
Concept ACK, a small comment :)

evanlinjin commented:
Concept ACK for me.

This PR is rather straight forward, and for the context on my work with multi-descriptor support (#486), it would be good to merge this PR soon (since I will also be modifying the `wallet/{mod|coin_selection}.rs` files).

Other than the concerns mentioned in https://github.com/bitcoindevkit/bdk/pull/662#discussion_r921181198, I think this PR is good to merge.

danielabrozzoni commented:
Please rebase to fix the conflicts, then I'll ACK :)

danielabrozzoni commented:
Code review ACK 6493bbf6db30c944c95321e359759126e8fa1350

csralvall commented:
Rebased with master.
- I've added an extra line to the `CHANGELOG.md` about the new meaning of
`fee_amount` in `CoinSelectionResult` which wasn't explicitly described before.
- I've removed the `Resolves: ...` part of the commit as I've realized it creates
a lot of automatic logs in related PRs and Issues, each time we add new changes,
duplicating info and making issue navigation cumbersome.
At the end, it's the maintainer who will put that line in the merge commit.
Maybe we can use some format to avoid the trigger and at the same time suggest
the fixed issues to the maintainer.
If you think it's not a serious issue, I can continue writing commits like
before.

danielabrozzoni commented on src/wallet/coin_selection.rs:
```diff
@@ -1505,12 +1480,11 @@ mod test {
             utxos,
             0,
             target_amount as i64,
-            FEE_AMOUNT,
```
In all the other tests you do `target_amount = *whatever it was before* + FEE_AMOUNT", maybe you should do the same here?

csralvall replied:
Sure! Although, now that I review the test again, it's impossible for it to fail :eyes:.

danielabrozzoni replied:
Yeah, you can't do much with unit testing on the SDR, other than hoping it doesn't panic

danielabrozzoni reviewed (approved):
tACK 6eb299b72f3126ae3d1e234b2a8039041273528f

danielabrozzoni commented:
I rebased by myself, now it's ready for merge

re-ACK e8df3d2d91927edb9a339c664f0603c47622e4b0 - I tested with the fuzzer, run it for 13,000,000 iterations, couldn't find any crash :)
